File: simple_lang_scala/simple_lang_parser/src/main/scala/simple_lang/tokenizer/Tokenizer.scala
package simple_lang.tokenizer

import scala.util.matching.Regex
import scala.util.parsing.combinator._

class TokenizerException(message: String) extends Exception(message)

object Tokenizer extends RegexParsers {
    
    def reserved_words: Parser[Token]  = {
        "if" ^^^ IfToken |
        "else" ^^^ ElseToken |
        "while" ^^^ WhileToken |
        "print" ^^^ PrintToken
    }
    
    def reserved_symbols: Parser[Token] = {
        "(" ^^^ LeftParenToken |
        ")" ^^^ RightParenToken |
        "{" ^^^ LeftCurlyBraceToken |
        "}" ^^^ RightCurlyBraceToken
    }
    
    def types: Parser[Token] = {
        "int" ^^^ IntToken |
        "bool" ^^^ BoolToken |
        "string" ^^^ StringToken
    }
    
    def binary_ops: Parser[Token] = {
        "+" ^^^ PlusToken |
        "-" ^^^ MinusToken |
        "<" ^^^ LessThanToken |
        ">" ^^^ GreaterThanToken |
        "==" ^^^ ExactlyEqualsToken |
        "!=" ^^^ NotEqualsToken
    }
    
    def unary_ops: Parser[Token] = {
        "!" ^^^ NotUnaryToken |
        "--" ^^^ MinusUnaryToken
    }
    
    def expressions: Parser[Token] = {
        "true" ^^^ TrueBooleanLiteralToken |
        "false" ^^^ FalseBooleanLiteralToken |
        "[0-9]+".r ^^ { case num => IntegerLiteralToken(num) } |
        """["][^"]*["]""".r ^^ { case string => StringLiteralToken(string) } |
        "[a-zA-Z_][a-zA-Z0-9_]*".r ^^ { case string => VariableToken(string) }
    }
    
    def tokens: Parser[List[Token]] = {
        phrase(rep(reserved_words | reserved_symbols | types | binary_ops | unary_ops | expressions))
    }
    
    def apply(code: String): Seq[Token] = {
        parse(tokens, code) match {
            case Success(result, _) => result
            case Failure(msg, _) => throw new TokenizerException(msg)
            case Error(msg, _) => throw new TokenizerException(msg)
        }
    }
}

File: simple_lang_java/simple-lang-parser-mvn/simple-lang-parser/src/main/java/com/simple_lang/tokenizer/Tokenizer.java
package simple_lang.tokenizer;

import java.util.*;
import java.util.regex.*;

public class Tokenizer {

    // Reserved words
    private static Token reservedWords(String code) {
        if (code.startsWith("if")) return new IfToken();
        if (code.startsWith("else")) return new ElseToken();
        if (code.startsWith("while")) return new WhileToken();
        if (code.startsWith("print")) return new PrintToken();
        return null;
    }

    // Reserved symbols
    private static Token reservedSymbols(String code) {
        if (code.startsWith("(")) return new LeftParenToken();
        if (code.startsWith(")")) return new RightParenToken();
        if (code.startsWith("{")) return new LeftCurlyBraceToken();
        if (code.startsWith("}")) return new RightCurlyBraceToken();
        if (code.startsWith("=")) return new AssignmentToken();
        if (code.startsWith(";")) return new SemicolonToken();
        return null;
    }

    // Types
    private static Token types(String code) {
        if (code.startsWith("int")) return new IntToken();
        if (code.startsWith("bool")) return new BoolToken();
        if (code.startsWith("string")) return new StringToken();
        return null;
    }

    // Binary operators
    private static Token binaryOps(String code) {
        if (code.startsWith("+")) return new PlusToken();
        if (code.startsWith("-")) return new MinusToken();
        if (code.startsWith("<")) return new LessThanToken();
        if (code.startsWith(">")) return new GreaterThanToken();
        if (code.startsWith("==")) return new ExactlyEqualsToken();
        if (code.startsWith("!=")) return new NotEqualsToken();
        return null;
    }

    // Unary operators
    private static Token unaryOps(String code) {
        if (code.startsWith("!")) return new NotUnaryToken();
        if (code.startsWith("--")) return new MinusUnaryToken();
        return null;
    }

     // Expressions
    private static Token expressions(String code) {
        //System.out.println("Current exp in expressions: " + code);
        if (code.startsWith("true")) return new TrueBooleanLiteralToken();
        if (code.startsWith("false")) return new FalseBooleanLiteralToken();
        
        Pattern p = Pattern.compile("[a-zA-Z_][a-zA-Z0-9_]*");
        Matcher m = p.matcher(code);
        if (m.find()) {
            String s = m.group();
            //System.out.println("We matched: " + s);
            return new VariableToken(s);
        }
        
        Pattern p1 = Pattern.compile("[0-9]+");
        Matcher m1 = p1.matcher(code);
        if (m1.find()) {
            String s1 = m1.group();
            //System.out.println("We matched: " + s1);
            return new IntegerLiteralToken(s1);
        }
        
        Pattern p2 = Pattern.compile("\"[^\"]*\"");
        Matcher m2 = p2.matcher(code);
        if (m2.find()) {
            String s2 = m2.group();
            //System.out.println("We matched: " + s2);
            return new StringLiteralToken(s2);
        }

        //Matcher integerMatcher = Pattern.compile("[0-9]+").matcher(code);
        //if (integerMatcher.find()) return new IntegerLiteralToken(integerMatcher.group());

        //Matcher stringMatcher = Pattern.compile("\"[^\"]*\"").matcher(code);
        //if (stringMatcher.find()) return new StringLiteralToken(stringMatcher.group());

        //Matcher variableMatcher = Pattern.compile("[a-zA-Z_][a-zA-Z0-9_]*").matcher(code);
        //if (variableMatcher.find()) return new VariableToken(variableMatcher.group());

        return null;
    }

    // Implementation of tokens in Tokenizer.scala
    public static List<Token> tokenize(String code) {
        List<Token> tokens = new ArrayList<>();
        int index = 0;

        // Iterates through the string to generate a list of tokens
        while (index < code.length()) {

            // Match each substring to a token type
            String substring = code.substring(index);

            //System.out.println("Remaining tokens: " + substring);
            //System.out.println("Current list: " + tokens.toString());
            
            Token token = reservedWords(substring);
            if (token != null) {
                //System.out.println("Successfully tokenized this reserved word: " + token.toString());
                tokens.add(token);
                index += token.getLength();
                continue;
            }

            token = reservedSymbols(substring);
            if (token != null) {
                //System.out.println("Successfully tokenized this symbol: " + token.toString());
                tokens.add(token);
                index += token.getLength();
                continue;
            }

            token = types(substring);
            if (token != null) {
                //System.out.println("Successfully tokenized this type: " + token.toString());
                tokens.add(token);
                index += token.getLength();
                continue;
            }

            token = binaryOps(substring);
            if (token != null) {
                //System.out.println("Successfully tokenized this binaryOp: " + token.toString());
                tokens.add(token);
                index += token.getLength();
                continue;
            }

            token = unaryOps(substring);
            if (token != null) {
                //System.out.println("Successfully tokenized this unaryOp: " + token.toString());
                tokens.add(token);
                index += token.getLength();
                continue;
            }

            token = expressions(substring);
            if (token != null) {
                //System.out.println("Successfully tokenized this exp: " + token.toString());
                //System.out.println("Token type variable???: " + (token instanceof VariableToken));
                tokens.add(token);
                index += token.getLength();
                continue;
            }

            // Token not found (does something else need to be done?)
            //Yes, we can either throw an exception here... or return a null list
            //Since it's already set up to check for a null list in apply, I'll do that
            //index++; removing
            tokens = null;
            return tokens;
        }
        return tokens;
    }

    // Implementation of apply method in Tokenizer.scala
    public static List<Token> apply(String code) throws TokenizerException {
        String noWhitespaceCode = code.replaceAll("\\s+", "");
        //System.out.println(noWhitespaceCode);
        List<Token> tokens = tokenize(noWhitespaceCode);
        if (tokens == null) {
            throw new TokenizerException("Failure");
        }
        return tokens;
    }
    
}
